### PR TITLE
Rewrite `put_start_and_end_on` to work in 3D

### PR DESCRIPTION
**Pull request by Darylgolden**
## Motivation
Previously,
```python
class MovingVertices(ThreeDScene):
    def construct(self):
        self.set_camera_orientation(phi=75 * DEGREES, theta=30 * DEGREES)
        vertices = [1, 2, 3, 4]
        edges = [(1, 2), (2, 3), (3, 4), (1, 3), (1, 4)]
        g = Graph(vertices, edges)
        self.play(Create(g))
        self.wait()
        self.play(g[1].animate.move_to([1, 1, 1]),
                  g[2].animate.move_to([-1, 1, 2]),
                  g[3].animate.move_to([1, -1, -1]),
                  g[4].animate.move_to([-1, -1, 0]))
        self.wait()
```
broke and rendered like this:

https://user-images.githubusercontent.com/9102529/115131773-377f6380-a02d-11eb-9c8c-a6e00153a270.mp4

Rewriting `put_start_and_end_on` to work in 3D fixes this:

https://user-images.githubusercontent.com/9102529/115131862-dc9a3c00-a02d-11eb-8adf-4fb4fa8574eb.mp4

the angle is slightly different? There seems to be non-deterministic output with that code. Stable also has that problem (at least on my system), so I don't think it's my fault.
## Further Comments
Dependent on #1354. I hope I did the git stuff properly.

## Checklist
- [ ] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [ ] I have written a descriptive PR title (see top of PR template for examples)
- [ ] I have written a changelog entry for the PR or deem it unnecessary
- [ ] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by Darylgolden** [comment]
This branch has the same non-deterministic errors I pointed out on #1354. I'm also not sure what to do about the failing tests on Mac, it seems like numpy evaluates the function slightly differently on there, causing a very small numerical error.

**Comment by jsonvillanueva** [comment]
> 
> 
> ~This branch has the same non-deterministic errors I pointed out on #1354.~ I'm also not sure what to do about the failing tests on Mac, it seems like numpy evaluates the function slightly differently on there, causing a very small numerical error.

Seems related to #972

**Comment by Darylgolden** [comment]
Even after rounding, the tests still fail on Mac because -0 is different from 0 for some reason.

**Comment by Darylgolden** [comment]
I managed to get the tests working, but using sort of hacky methods.

**Comment by Darylgolden** [comment]
I've written a test in three_d, let me know if it's misplaced.

File: tests/test_graphical_units/test_threed.py [hunk 1]
```diff
@@ -76,6 +76,22 @@ def construct(self):
         self.add(axes)
 
 
+class MovingVerticesTest(ThreeDScene):
+    def construct(self):
+        self.set_camera_orientation(phi=75 * DEGREES, theta=30 * DEGREES)
+        vertices = [1, 2, 3, 4]
+        edges = [(1, 2), (2, 3), (3, 4), (1, 3), (1, 4)]
+        g = Graph(vertices, edges)
```
Comment (behackl) [inline tests/test_graphical_units/test_threed.py:84]: It's not really important, but I would probably specify `vertex_type=Dot3D` if the vertices actually have 3d coordinates.

**Review by behackl — APPROVED** [review]
LGTM! I only have one cosmetic remark (which doesn't really matter though); I think this can also be merged as-is.

File: tests/test_graphical_units/test_threed.py (hunk 1 above)
Comment (Darylgolden) [reply tests/test_graphical_units/test_threed.py]: I've changed the vertex type.

**Comment by behackl** [comment]
I noticed that the test time went up by a bit, and after checking the test out locally I see that my suggestion was bad: Dot3D is rather heavy for the rendering (although it looks better in the example). I've reverted the test and the corresponding data back to using the default vertex type. If the pipeline passes after this, the PR can be merged.

**Pull request merged**